File: web/app/announcements.js
// Announcements are single files that play instead of the music at their
// own times: a message for customers, an offer, an advertisement. The panel
// lists them and holds the editor.

import * as A from './api.js';
import { h, clear, icon, toast, confirmDialog, modal, spinner, errorBox } from './dom.js';
import { libraryPicker, fail, baseOf } from './fileops.js';

// The week starts on Sunday, as everywhere else. The bits are the API's:
// Monday is 1, Sunday is 64.
const DAYS = [['Sun', 64], ['Mon', 1], ['Tue', 2], ['Wed', 4], ['Thu', 8], ['Fri', 16], ['Sat', 32]];

const KINDS = [
  ['file', 'This file', 'The same file plays every time.'],
  ['random', 'Any file of a folder', 'One file of the folder plays, chosen at random.'],
  ['cycle', 'The files of a folder in turn', 'The next file of the folder plays each time.'],
];

function dayLabel(mask) {
  if (mask === 127) return 'Every day';
  if (mask === 31) return 'Mon–Fri';
  if (mask === 96) return 'Weekends';
  return DAYS.filter(([, b]) => mask & b).map(([l]) => l).join(' ');
}

// summary is the one line under the name of an announcement.
function summary(a) {
  const when = a.mode === 'at'
    ? `at ${a.at_time}`
    : `every ${a.every_minutes} min from ${a.start_time} to ${a.end_time}`;
  const what = a.source_kind === 'file' ? baseOf(a.source_ref) : `${a.source_kind === 'random' ? 'any file' : 'files in turn'} of /${a.source_ref || ''}`;
  return `${dayLabel(a.days)} · ${when} · ${what}${a.volume != null ? ` · vol ${a.volume}` : ''}`;
}

// announcementsPanel returns {el, load, onEvent, destroy}.
export function announcementsPanel() {
  const body = h('div');
  const head = h('div.panel-head', h('h2.panel-title', 'Announcements'),
    h('div.tools', h('button.btn.btn-sm.btn-outline-secondary', { type: 'button', onclick: () => editor(null) }, icon('plus-lg'), 'Add announcement')));
  const el = h('div.panel.clip', head, body);

  async function load() {
    clear(body).append(spinner());
    let list;
    try { list = (await A.api.get('/announcements')).announcements; } catch (e) { clear(body).append(h('div.panel-body', errorBox(e))); return; }
    clear(body);
    if (!list.length) {
      body.append(h('div.panel-empty', 'No announcements. Add one to play a message, an offer or an advertisement between the music.'));
      return;
    }
    const rows = h('div.rows');
    for (const a of list) {
      const sw = h('input.form-check-input', { type: 'checkbox', role: 'switch', checked: a.enabled, 'aria-label': `Enable ${a.name}`, onchange: async () => {
        try { await A.api.put(`/announcements/${a.id}`, { ...a, enabled: sw.checked }); } catch (err) { sw.checked = !sw.checked; fail(err); }
      } });
      rows.append(h('div.row-item.tall',
        h('div.form-check.form-switch.switch-only', sw),
        h('div.row-main', h('div.row-title.fw-semibold', a.name), h('div.row-sub.text-wrap', summary(a))),
        h('button.btn.btn-outline-secondary.btn-icon', { type: 'button', 'aria-label': `Play ${a.name} now`, title: 'Play it now. The music continues afterwards.', onclick: () => playNow(a) }, icon('megaphone')),
        h('button.btn.btn-outline-secondary.btn-icon', { type: 'button', 'aria-label': `Edit ${a.name}`, title: 'Edit', onclick: () => editor(a) }, icon('pencil')),
        h('button.btn.btn-outline-danger.btn-icon', { type: 'button', 'aria-label': `Delete ${a.name}`, title: 'Delete', onclick: async () => {
          if (!await confirmDialog({ title: 'Delete announcement', body: `Delete "${a.name}"?`, confirmText: 'Delete', danger: true })) return;
          try { await A.api.del(`/announcements/${a.id}`); load(); } catch (err) { fail(err); }
        } }, icon('trash'))));
    }
    body.append(rows);
  }

  async function playNow(a) {
    try {
      await A.api.post(`/announcements/${a.id}/play`);
      toast(`Playing ${a.name}`, 'success');
    } catch (err) { fail(err); }
  }

  function editor(ann) {
    const init = ann || { name: '', enabled: true, days: 127, mode: 'at', at_time: '10:15', start_time: '09:00', end_time: '17:00', every_minutes: 30, source_kind: 'file', source_ref: '', volume: null };
    const name = h('input.form-control', { type: 'text', value: init.name, required: true, maxlength: 100, placeholder: 'Closing time message' });
    const dayChecks = DAYS.map(([label, bit]) => {
      const c = h('input.btn-check', { type: 'checkbox', id: `ann-day-${bit}`, checked: !!(init.days & bit), autocomplete: 'off' });
      return [c, h('label.btn.btn-sm', { for: `ann-day-${bit}` }, label)];
    });

    const mode = h('select.form-select',
      h('option', { value: 'at', selected: init.mode !== 'every' }, 'Once a day, at a time'),
      h('option', { value: 'every', selected: init.mode === 'every' }, 'Again and again, between two times'));
    const atTime = h('input.form-control', { type: 'time', value: init.at_time || '10:15' });
    const startTime = h('input.form-control', { type: 'time', value: init.start_time || '09:00' });
    const endTime = h('input.form-control', { type: 'time', value: init.end_time || '17:00' });
    const every = h('input.form-control', { type: 'number', min: 1, max: 1440, value: init.every_minutes ?? 30 });
    const atRow = h('div.mt-2', h('label.form-label', 'Time'), atTime);
    const everyRow = h('div.row.g-2.mt-2',
      h('div.col-4', h('label.form-label', 'From'), startTime),
      h('div.col-4', h('label.form-label', 'To'), endTime),
      h('div.col-4', h('label.form-label', 'Every (min)'), every));

    const kind = h('select.form-select');
    for (const [value, label] of KINDS) kind.append(h('option', { value, selected: init.source_kind === value }, label));
    const kindNote = h('div.form-text');
    const ref = h('input.form-control.mono', { type: 'text', value: init.source_ref || '', placeholder: 'announcements/closing.mp3', 'aria-label': 'File or folder' });
    const browse = h('button.btn.btn-outline-secondary', { type: 'button', onclick: () => {
      if (kind.value === 'file') {
        libraryPicker({ title: 'Choose the file', files: true, onPick: (files) => { if (files.length) ref.value = files[0]; } });
      } else {
        libraryPicker({ title: 'Choose the folder', start: ref.value, onPick: (p) => { ref.value = p; } });
      }
    } }, 'Browse…');

    const useVol = h('input', { type: 'checkbox', checked: init.volume != null });
    const vol = h('input.form-range.mt-1', { type: 'range', min: 0, max: 100, value: init.volume ?? 80, disabled: init.volume == null, 'aria-label': 'Announcement volume' });
    const volLabel = h('span.mono', String(init.volume ?? 80));
    vol.addEventListener('input', () => { volLabel.textContent = vol.value; });
    useVol.addEventListener('change', () => { vol.disabled = !useVol.checked; });

    const sync = () => {
      atRow.classList.toggle('hidden', mode.value !== 'at');
      everyRow.classList.toggle('hidden', mode.value !== 'every');
      ref.placeholder = kind.value === 'file' ? 'announcements/closing.mp3' : 'announcements';
      kindNote.textContent = KINDS.find(([v]) => v === kind.value)[2];
    };
    mode.addEventListener('change', sync);
    kind.addEventListener('change', sync);
    sync();

    const err = h('div.form-error');
    const form = h('form', { onsubmit: async (e) => {
      e.preventDefault();
      const days = dayChecks.reduce((m, [c], i) => m | (c.checked ? DAYS[i][1] : 0), 0);
      if (!days) { err.textContent = 'Choose at least one day.'; return; }
      const value = ref.value.trim().replace(/^\/+|\/+$/g, '');
      if (kind.value === 'file' && !value) { err.textContent = 'Choose the file to play.'; return; }
      if (kind.value !== 'file' && !value) { err.textContent = 'Choose the folder to play from.'; return; }
      const body = {
        name: name.value.trim(), enabled: init.enabled, days, mode: mode.value,
        source_kind: kind.value, source_ref: value,
        volume: useVol.checked ? Number(vol.value) : null,
      };
      if (mode.value === 'at') body.at_time = atTime.value;
      else Object.assign(body, { start_time: startTime.value, end_time: endTime.value, every_minutes: Number(every.value) });
      try {
        if (ann) await A.api.put(`/announcements/${ann.id}`, body);
        else await A.api.post('/announcements', body);
        dlg.hide();
        load();
      } catch (ex) { err.textContent = ex.message; }
    } },
      h('label.form-label', 'Name'), name,
      h('label.form-label.mt-3', 'Days'),
      h('div.d-flex.flex-wrap.gap-1', dayChecks.flat()),
      h('label.form-label.mt-3', 'When'), mode, atRow, everyRow,
      h('label.form-label.mt-3', 'What plays'), kind, kindNote,
      h('div.input-row.mt-2', ref, browse),
      h('label.check-line', useVol, 'Play it at its own volume:', volLabel), vol,
      h('div.form-text', 'The music stops for the announcement and starts again where it was.'),
      err);
    const dlg = modal({
      title: ann ? 'Edit announcement' : 'New announcement',
      body: form,
      footer: [
        h('button.btn.btn-outline-secondary', { type: 'button', 'data-bs-dismiss': 'modal' }, 'Cancel'),
        h('button.btn.btn-primary', { type: 'button', onclick: () => form.requestSubmit() }, 'Save'),
      ],
    });
  }

  return { el, load, onEvent(type) { if (type === 'schedule') load(); }, destroy() {} };
}

File: internal/store/announcements.go
package store

import (
	"context"
	"database/sql"
	"errors"
	"time"
)

// Announcement is one file that plays instead of the music at its own
// times. The days use the same bits as Schedule.
type Announcement struct {
	ID           int64      `json:"id,omitempty" doc:"Set by the server"`
	Name         string     `json:"name" minLength:"1" maxLength:"100"`
	Enabled      bool       `json:"enabled"`
	Days         int        `json:"days" minimum:"1" maximum:"127" doc:"Bit mask, Monday is 1, Sunday is 64"`
	Mode         string     `json:"mode" enum:"at,every" doc:"at plays once a day at at_time. every plays from start_time to end_time, one play every every_minutes."`
	AtTime       *string    `json:"at_time,omitempty" pattern:"^([01][0-9]|2[0-3]):[0-5][0-9]$" doc:"Local wall-clock time HH:MM, for mode at"`
	StartTime    *string    `json:"start_time,omitempty" pattern:"^([01][0-9]|2[0-3]):[0-5][0-9]$" doc:"First play, for mode every"`
	EndTime      *string    `json:"end_time,omitempty" pattern:"^([01][0-9]|2[0-3]):[0-5][0-9]$" doc:"No play after this time, for mode every"`
	EveryMinutes *int       `json:"every_minutes,omitempty" minimum:"1" maximum:"1440" doc:"Minutes between plays, for mode every"`
	SourceKind   string     `json:"source_kind" enum:"file,random,cycle" doc:"file plays source_ref. random plays one file of the folder source_ref. cycle plays the next file of that folder each time."`
	SourceRef    string     `json:"source_ref" doc:"A file or a folder under the music root"`
	CycleIndex   int        `json:"-" doc:"Position in the folder, for cycle"`
	Volume       *int       `json:"volume,omitempty" minimum:"0" maximum:"100" doc:"Volume for the announcement. The music volume returns after it."`
	LastPlayed   *time.Time `json:"last_played,omitempty" doc:"When it played last"`
	CreatedAt    time.Time  `json:"created_at,omitempty"`
}

const announcementColumns = `id, name, enabled, days, mode, at_time, start_time, end_time, every_minutes,
	source_kind, source_ref, cycle_index, volume, last_played, created_at`

func scanAnnouncement(sc interface{ Scan(...any) error }) (Announcement, error) {
	var a Announcement
	var at, start, end, last sql.NullString
	var every, vol sql.NullInt64
	var created string
	err := sc.Scan(&a.ID, &a.Name, &a.Enabled, &a.Days, &a.Mode, &at, &start, &end, &every,
		&a.SourceKind, &a.SourceRef, &a.CycleIndex, &vol, &last, &created)
	if err != nil {
		return a, err
	}
	if at.Valid {
		a.AtTime = &at.String
	}
	if start.Valid {
		a.StartTime = &start.String
	}
	if end.Valid {
		a.EndTime = &end.String
	}
	if every.Valid {
		v := int(every.Int64)
		a.EveryMinutes = &v
	}
	if vol.Valid {
		v := int(vol.Int64)
		a.Volume = &v
	}
	if last.Valid {
		t := parse(last.String)
		a.LastPlayed = &t
	}
	a.CreatedAt = parse(created)
	return a, nil
}

// ListAnnouncements returns every announcement by name.
func (s *Store) ListAnnouncements(ctx context.Context) ([]Announcement, error) {
	rows, err := s.r.QueryContext(ctx, `SELECT `+announcementColumns+` FROM announcements ORDER BY name COLLATE NOCASE`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	out := []Announcement{}
	for rows.Next() {
		a, err := scanAnnouncement(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, a)
	}
	return out, rows.Err()
}

// GetAnnouncement returns one announcement.
func (s *Store) GetAnnouncement(ctx context.Context, id int64) (Announcement, bool, error) {
	row := s.r.QueryRowContext(ctx, `SELECT `+announcementColumns+` FROM announcements WHERE id = ?`, id)
	a, err := scanAnnouncement(row)
	if errors.Is(err, sql.ErrNoRows) {
		return a, false, nil
	}
	if err != nil {
		return a, false, err
	}
	return a, true, nil
}

// CreateAnnouncement inserts one and returns it with its id.
func (s *Store) CreateAnnouncement(ctx context.Context, a Announcement) (Announcement, error) {
	a.CreatedAt = time.Now().UTC()
	res, err := s.w.ExecContext(ctx, `INSERT INTO announcements
		(name, enabled, days, mode, at_time, start_time, end_time, every_minutes, source_kind, source_ref, cycle_index, volume, created_at)
		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, 0, ?, ?)`,
		a.Name, a.Enabled, a.Days, a.Mode, nullString(a.AtTime), nullString(a.StartTime), nullString(a.EndTime),
		nullInt(a.EveryMinutes), a.SourceKind, a.SourceRef, nullInt(a.Volume), format(a.CreatedAt))
	if err != nil {
		return a, err
	}
	a.ID, err = res.LastInsertId()
	return a, err
}

// UpdateAnnouncement replaces one. The cycle position and the last play
// stay as they are.
func (s *Store) UpdateAnnouncement(ctx context.Context, a Announcement) error {
	_, err := s.w.ExecContext(ctx, `UPDATE announcements SET name = ?, enabled = ?, days = ?, mode = ?, at_time = ?,
		start_time = ?, end_time = ?, every_minutes = ?, source_kind = ?, source_ref = ?, volume = ? WHERE id = ?`,
		a.Name, a.Enabled, a.Days, a.Mode, nullString(a.AtTime), nullString(a.StartTime), nullString(a.EndTime),
		nullInt(a.EveryMinutes), a.SourceKind, a.SourceRef, nullInt(a.Volume), a.ID)
	return err
}

// DeleteAnnouncement removes one.
func (s *Store) DeleteAnnouncement(ctx context.Context, id int64) error {
	_, err := s.w.ExecContext(ctx, `DELETE FROM announcements WHERE id = ?`, id)
	return err
}

// MarkAnnouncementPlayed records the play time and the next position of a
// cycle, so a restart does not repeat the same file.
func (s *Store) MarkAnnouncementPlayed(ctx context.Context, id int64, at time.Time, cycleIndex int) error {
	_, err := s.w.ExecContext(ctx, `UPDATE announcements SET last_played = ?, cycle_index = ? WHERE id = ?`,
		format(at.UTC()), cycleIndex, id)
	return err
}

func nullString(s *string) any {
	if s == nil {
		return nil
	}
	return *s
}

func nullInt(v *int) any {
	if v == nil {
		return nil
	}
	return *v
}
